private/buf/bufcli/bufcli.go

```diff
@@ -725,6 +738,103 @@ func ReadModuleWithWorkspacesDisabled(
 	return module, moduleIdentity, err
 }
 
+// NewImageForSource acts like the 'buf build' command, such that it resolves a single
+// bufimage.Image from the user-provided source. This function is primarily used by the
+// 'buf encode' and 'buf decode' commands, so it doesn't need to support all of the build
+// options, such as paths and exclude paths.
+func NewImageForSource(
+	ctx context.Context,
+	container appflag.Container,
+	registryProvider registryv1alpha1apiclient.Provider,
+	source string,
+	errorFormat string,
+) (bufimage.Image, error) {
+	ref, err := buffetch.NewRefParser(container.Logger(), buffetch.RefParserWithProtoFileRefAllowed()).GetRef(ctx, source)
+	if err != nil {
+		return nil, err
+	}
+	storageosProvider := storageos.NewProvider(storageos.ProviderWithSymlinks())
+	imageConfigReader, err := NewWireImageConfigReader(
+		container,
+		storageosProvider,
+		command.NewRunner(),
+		registryProvider,
+	)
+	if err != nil {
+		return nil, err
+	}
+	imageConfigs, fileAnnotations, err := imageConfigReader.GetImageConfigs(
+		ctx,
+		container,
+		ref,
+		"",
+		nil,
+		nil,
+		false,
+		false,
+	)
+	if err != nil {
+		return nil, err
+	}
+	if len(fileAnnotations) > 0 {
+		if err := bufanalysis.PrintFileAnnotations(
+			container.Stderr(),
+			fileAnnotations,
+			errorFormat,
+		); err != nil {
+			return nil, err
+		}
+		return nil, ErrFileAnnotation
+	}
+	images := make([]bufimage.Image, 0, len(imageConfigs))
+	for _, imageConfig := range imageConfigs {
+		images = append(images, imageConfig.Image())
+	}
+	return bufimage.MergeImages(images...)
+}
+
+// NewDescriptorBytesForInput returns the descriptor bytes for an input, which can be
+// from stdin, argument, or a file.
+func NewDescriptorBytesForInput(
+	ctx context.Context,
+	container appflag.Container,
+	registryProvider registryv1alpha1apiclient.Provider,
+	input string,
+) ([]byte, error) {
+	ref, err := buffetch.NewRefParser(container.Logger(), buffetch.RefParserWithProtoFileRefAllowed()).GetRef(ctx, input)
+	if err != nil {
+		return nil, err
+	}
+	switch t := ref.(type) {
+	case buffetch.SourceRef:
+		return []byte(input), nil
+	case buffetch.ImageRef:
```

Comment:
Similar to the above, this is actually abusing the `.bin` extension a bit since that's reserved for Image files. But this isn't actually an Image (which is a `FileDescriptorSet`) - this is a binary-encoded Protobuf message.

All of this works because we're able to use `GetImageFile` and that just reads the content directly. The rest of the `moduleResolver`, etc. aren't actually ever used.

Again, it's probably easiest just to require that the user provides the bytes on `stdin`, and we leave it at that for now.